### PR TITLE
iteration documentation and limitations based on unit tests first draft

### PR DESCRIPTION
**Pull request by moAlobaidi**
Hey @collinc97 , here's a first draft of the documentation of loops and iteration in Leo based on the unit tests you gave me for reference. Lmk if any of this information is not accurate or if there is anything else I should add! Also feel free to give any feedback necessary on the style / wording. 

This PR resolves issue #56

### CODE REVIEW

File: documentation/leo/03_language.md
Comment (collinc97) [inline documentation/leo/03_language.md:535]: make these more specific: say exactly what the minimum and maximum loop requirements are or remove these docs

**Review by collinc97 — CHANGES_REQUESTED** [review]
Thanks for the PR @moAlobaidi!
Definitely looks like you spent some time playing around with loops. An overall comment from me would be to follow the length and diction of the other Leo docs.

File: documentation/leo/03_language.md
Comment (moAlobaidi) [reply documentation/leo/03_language.md:496]: Good point! Would an alternate implementation for non-literal loop bounds be asserting that the bound isn't too large?

File: documentation/leo/03_language.md
Comment (aharshbe) [inline documentation/leo/03_language.md:435]: For consistency with the existing docs, let's change <> to {}, so `for {variable} in {lower-bound}..{upper-bound}`. A good example of this to reference is the [mapping type](https://developer.aleo.org/leo/language#mapping).

File: documentation/leo/03_language.md
Comment (aharshbe) [inline documentation/leo/03_language.md:441]: Same thing here change <> to {}

File: documentation/leo/03_language.md
Comment (aharshbe) [inline documentation/leo/03_language.md:437]: After this sentence, add "Nested loops are supported." and remove the example of a nested loop below.

File: documentation/leo/03_language.md
Comment (aharshbe) [inline documentation/leo/03_language.md:447]: I don't think you need to provide an example here. Just say nested loops are supported.

**Review by aharshbe — CHANGES_REQUESTED** [review]
I left a few simple changes here. Happy to make them myself but want to make sure you have some eyes on it first @moAlobaidi 🙂 Nice work!

**Comment by aharshbe** [comment]
Once these changes are complete, this is good to merge.

**Comment by aharshbe** [comment]
Let's hold on merging this until we confirm u64 vs u32 with Pranav tomorrow.

File: documentation/leo/03_language.md
Comment (aharshbe) [inline documentation/leo/03_language.md:436]: @moAlobaidi, this should change to support up to u32. Just got confirmation from the Leo team. For context, internally we support up to u128 but externally we shouldn't mention this.

**Review by aharshbe — APPROVED** [review]
@moAlobaidi, we just need to update to be up to u32 so remove u64. Going to approve this so once you made those changes feel free to merge!

cc @collinc97

**Review by aharshbe — APPROVED** [review]

File: documentation/leo/03_language.md
Comment (collinc97) [inline documentation/leo/03_language.md:436]:
```suggestion
For Loops are declared as `for {loop variable} in {lower bound}..{upper bound}`.  Typical Leo loop variable types include `u8`, `u16`, and `u32`. The lower bound must be
```

Comment (collinc97) [reply documentation/leo/03_language.md:436]: What is a `{loop variable}` it is not defined in the documentation.

File: documentation/leo/03_language.md
Comment (collinc97) [inline documentation/leo/03_language.md:444]: You should add a code snippet of a loop that actually compiles.
People will want to copy and paste code from the docs into their program to play around with the feature.

**Review by collinc97 — CHANGES_REQUESTED** [review]

File: documentation/leo/03_language.md
Comment (collinc97) [reply documentation/leo/03_language.md]: That description still does not help the programmer write a loop @moAlobaidi.
Loops in Leo require a variable name and a variable type.
`{loop variable}` should be defined as `{name: type}` with an example.

Comment (moAlobaidi) [reply documentation/leo/03_language.md]: Okay updated.

File: documentation/leo/03_language.md
Comment (collinc97) [inline documentation/leo/03_language.md:459]:
```suggestion
For Loops are declared as `for {variable: type} in {lower bound: type}..{upper bound: type}`. Unsigned integer 
```
Remove mixed usage of snake_case and text in description

File: documentation/leo/03_language.md
Comment (collinc97) [inline documentation/leo/03_language.md:461]:
```suggestion
types `u8`, `u16`, and `u32` are recommended for loop variable types. The lower bound must be
```
* State explicit types.
* Add backticks to types so they show up as `code` - this is consistent with the rest of `u8` mentions throughout the documentation.
* Remove usage of "iterator" name which has not been mentioned before in documentation
* Specify that we are describing `type` of. the iterator variable.

File: documentation/leo/03_language.md
Comment (collinc97) [inline documentation/leo/03_language.md:469]:
```suggestion
  for i: u32 in 0u32..5u32 {
```

**Review by collinc97 — CHANGES_REQUESTED** [review]
final changes

File: documentation/leo/03_language.md
Comment (collinc97) [inline documentation/leo/03_language.md:459]:
```suggestion
For Loops are declared as `for {variable: type} in {lower bound}..{upper bound}`. Unsigned integer 
```
The lower and upper bounds are literals - their type is specified with the value `1u8`, `1337u16`
There is no colon required @moAlobaidi

**Review by collinc97 — CHANGES_REQUESTED** [review]

File: documentation/leo/03_language.md
Comment (collinc97) [inline documentation/leo/03_language.md:472]:
```suggestion
  return count; // returns 5u32
```

**Pull request merged**
